hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -51,13 +66,57 @@ public TokenRejectHandler() {}
     @Override
     public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
         requireNonNull(context);
-        // Todo implement the preHandle logic
+        final var txn = context.body();
+        final var op = txn.tokenRejectOrThrow();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+        var senderAccount = op.account();
+        verifySenderAndRequireKey(senderAccount, context, accountStore);
     }
 
+    private void verifySenderAndRequireKey(
+            final AccountID senderId, final PreHandleContext context, final ReadableAccountStore accountStore)
+            throws PreCheckException {
+
+        final var senderAccount = accountStore.getAliasedAccountById(senderId);
+        if (senderAccount == null) {
+            throw new PreCheckException(INVALID_ACCOUNT_ID);
+        }
+
+        // If the sender account is immutable, then we throw an exception.
+        final var key = senderAccount.key();
+        if (key == null || !isValid(key)) {
+            throw new PreCheckException(ACCOUNT_IS_IMMUTABLE);
+        }
+        context.requireKey(key);
+    }
+
+    @SuppressWarnings("java:S2259")
     @Override
     public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
-        // Todo: Implement the pureChecks logic
-        throw new UnsupportedOperationException("Not implemented yet.");
+        requireNonNull(txn, "Transaction body cannot be null");
+        final var op = txn.tokenRejectOrThrow();
+
+        validateTruePreCheck(op.hasAccount(), INVALID_ACCOUNT_ID);
+        validateFalsePreCheck(op.rejections().isEmpty(), INVALID_TRANSACTION_BODY);
+
+        var uniqueTokenReferences = new HashSet<TokenReference>();
+        for (final var rejection : op.rejections()) {
+            if (!uniqueTokenReferences.add(rejection)) {
+                throw new PreCheckException(TOKEN_ID_REPEATED_IN_TOKEN_LIST);
```

Comment:
Need some opinions on this status. We don't have TOKEN_LIST here, we have TokenReferences as the new type introduced with the protobuf changes for TokenReject functionality.

A TokenReference may not always refer to TOKEN_ID, but also to NFT_ID.

I found this status as the most descriptive for this case.
We could also throw INVALID_TRANSACTION_BODY, but it won't be as clear what is the issue.

Reply:
Right, so since we introduced TokenReference as a type we might need new response codes related to it.
In this instance do you think `TOKEN_REFERENCE_REPEATED` is clear enough?

Reply:
Indeed, `TOKEN_REFERENCE_REPEATED` clearly describes the problem

Reply:
Added in https://github.com/hashgraph/hedera-protobufs/pull/341.